### PR TITLE
Pipeline 프로세서 모방해보기 in Rust (Step 1 ~ 3)

### PR DESCRIPTION
**Pull request by sigridjineth**
## Changes Made
* The processor initialization with the simple operation to pipelining
* Add data hazard cases

### CODE REVIEW

**Review by y123ob — APPROVED** [review]
3개 파이프라인 (Fetch, Decode, Execute) 를 만들고 instruction2 가 1의 dependency를 가져 2가 죽는hazard가 보이네요. LGTM

File: sigridjineth/ch4/README.md [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+  Executing instruction: instr2
+  ```
+  In the third cycle, both instructions are executed.
+
+- **4th Clock Cycle**:
+  ```
+  Fetching instruction: instr1
+  Fetching instruction: instr2
+  ```
+  After execution, the instructions are looped back to the Fetch stage. This isn't how real CPUs work (once an instruction is executed, it would be removed from the pipeline), but it's what's currently implemented in your simulation.
```
Comment (mconcat) [inline sigridjineth/ch4/README.md:69]: In real world, the instruction counter advances to the consequent memory address, so the pipelining cycle is applied on the next instructions async?

File: sigridjineth/ch4/README.md [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+
+This cycle of `Fetch -> Decode -> Execute -> Fetch` continues for each instruction until the 10 clock cycles are completed, which is why you see the repeated pattern in your output.
+
+## Step 2
+### Changes Made
```
Comment (mconcat) [inline sigridjineth/ch4/README.md:74]: I see that the program counter advancement happens in this step, good work!

File: sigridjineth/ch4/README.md [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+Decoding instruction: instr1
+Cycle: 7
+Completed instruction: instr1
+Cycle: 8
+Cycle: 9
+```
+
+## Step 3: Data Hazards
+### What is this?
+The concept of parallel execution can introduce potential risks or hazards when instructions depend on the results of other instructions.
```
Comment (mconcat) [inline sigridjineth/ch4/README.md:107]: I was wondering how would pipeline detects the data hazard while reading JBs PR. Reading the code, it seems like that each instruction gets assigned with what register/memory it depends on, and the data hazard detector(idk if this is the name) checks the dependency?

File: sigridjineth/ch4/README.md (hunk 1 above)
Comment (sigridjineth) [reply sigridjineth/ch4/README.md:69]: Yes, the pipeline has been done asynchronously in the sense that each stage of the pipeline operates independently and concurrently. When there is no hazard, each clock cycle will see a new instruction enter the fetch stage, while other instructions advance to decode, execute, etc., making full use of the pipeline's parallelism.

File: sigridjineth/ch4/README.md (hunk 3 above)
Comment (sigridjineth) [reply sigridjineth/ch4/README.md:107]: Look: https://blog.naver.com/PostView.naver?blogId=mayooyos&logNo=220878015447

**Comment by sigridjineth** [comment]
https://courses.cs.washington.edu/courses/cse378/02sp/sections/section7-1.html

**Pull request merged**